File: src/10.graph/05.401.go
package graph

import (
	"fmt"
	"math/bits"
)

// ReadBinaryWatch 枚举法
//
//	@param turnedOn
//	@return []string
func ReadBinaryWatch1(turnedOn int) []string {

	ans := []string{}
	for h := 0; h < 12; h++ {
		for m := 0; m < 60; m++ {

			if bits.OnesCount(uint(h))+bits.OnesCount(uint(m)) == turnedOn {
				ans = append(ans, fmt.Sprintf("%d:%02d", h, m))
			}
		}
	}
	return ans
}

// ReadBinaryWatch 回溯法
//
//	@param turnedOn
//	@return []string
func ReadBinaryWatch(turnedOn int) []string {
	var backtrack func(h, m, num int, hour, minutes []int, isHour bool)
	ans := []string{}
	backtrack = func(h, m, num int, hour, minutes []int, isHour bool) {
		if num == 0 {
			ans = append(ans, fmt.Sprintf("%d:%02d", h, m))
			return
		}

		if isHour {
			for i, v := range hour {
				if h+v < 12 {
					backtrack(h+v, m, num-1, hour[i+1:], minutes, true)
				}
			}
		}

		for i, v := range minutes {
			if m+v < 60 {
				backtrack(h, m+v, num-1, hour, minutes[i+1:], false)
			}
		}
	}

	backtrack(0, 0, turnedOn, []int{1, 2, 4, 8}, []int{1, 2, 4, 8, 16, 32}, true)

	return ans
}

File: src/10.graph/05.401.ts
/**
 * 二进制枚举法
 * @param turnedOn
 * @returns
 */
export function readBinaryWatch1(turnedOn: number): string[] {
    const ans: string[] = [];
    for (let h = 0; h < 12; h++) {
        for (let m = 0; m < 60; m++) {
            // h.toString(2) 转为 2 进制
            // 判断该 2 进制数中有几个 1
            if (
                h.toString(2).split("0").join("").length +
                    m.toString(2).split("0").join("").length ===
                turnedOn
            ) {
                ans.push(h + ":" + (m < 10 ? "0" : "") + m);
            }
        }
    }
    return ans;
}

/**
 * 回溯法
 * @param turnedOn
 * @returns
 */
export function readBinaryWatch(turnedOn: number): string[] {
    const ans: string[] = [];

    const backtrack = (
        h: number,
        m: number,
        num: number,
        hour: number[],
        minutes: number[],
        isHour: Boolean
    ) => {
        if (num === 0) {
            ans.push(h + ":" + (m < 10 ? "0" : "") + m);
            return;
        }

        if (isHour) {
            hour.forEach((_h, i) => {
                if (h + _h < 12) {
                    backtrack(
                        h + _h,
                        m,
                        num - 1,
                        hour.slice(i + 1),
                        minutes,
                        true
                    );
                }
            });
        }

        minutes.forEach((_m, i) => {
            if (m + _m < 60) {
                backtrack(
                    h,
                    m + _m,
                    num - 1,
                    hour,
                    minutes.slice(i + 1),
                    false
                );
            }
        });
    };

    backtrack(0, 0, turnedOn, [1, 2, 4, 8], [1, 2, 4, 8, 16, 32], true);

    return ans;
}
